pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -338,55 +338,76 @@ func (config *V2RegistriesConf) postProcess() error {
-		return ctx.SystemRegistriesConfDirPath
+	if newConfigWrapper(ctx).userConfigDirPath != "" {
+		return newConfigWrapper(ctx).userConfigDirPath
 	}
-
-	userRegistriesDirPath := filepath.Join(homedir.Get(), userRegistriesDir)
-	if _, err := os.Stat(userRegistriesDirPath); err == nil {
-		return userRegistriesDirPath
-	}
-
-	if ctx != nil && ctx.RootForImplicitAbsolutePaths != "" {
-		return filepath.Join(ctx.RootForImplicitAbsolutePaths, systemRegistriesConfDirPath)
-	}
-
-	return systemRegistriesConfDirPath
+	return newConfigWrapper(ctx).configDirPath
 }
 
 // configWrapper is used to store the paths from ConfigPath and ConfigDirPath
 // and acts as a key to the internal cache.
 type configWrapper struct {
-	configPath    string
+	// path to the registries.conf file
+	configPath string
+	// path to system-wide registries.conf.d directory, or "" if not used
 	configDirPath string
+	// path to user specificed registries.conf.d directory, or "" if not used
+	userConfigDirPath string
 }
 
 // newConfigWrapper returns a configWrapper for the specified SystemContext.
 func newConfigWrapper(ctx *types.SystemContext) configWrapper {
-	return configWrapper{
-		configPath:    ConfigPath(ctx),
-		configDirPath: ConfigDirPath(ctx),
+	var wrapper configWrapper
+	userRegistriesFilePath := filepath.Join(homedir.Get(), userRegistriesFile)
+	userRegistriesDirPath := filepath.Join(homedir.Get(), userRegistriesDir)
+	if ctx == nil {
+		wrapper.configPath = systemRegistriesConfPath
+	} else {
+		if ctx.SystemRegistriesConfPath != "" {
+			wrapper.configPath = ctx.SystemRegistriesConfPath
+		} else if _, err := os.Stat(userRegistriesFilePath); err == nil {
+			wrapper.configPath = userRegistriesFilePath
+			if ctx.SystemRegistriesConfDirPath != "" {
+				wrapper.configDirPath = ctx.SystemRegistriesConfDirPath
+			} else {
+				wrapper.userConfigDirPath = userRegistriesDirPath
+			}
+			return wrapper
+		} else if ctx.RootForImplicitAbsolutePaths != "" {
+			wrapper.configPath = filepath.Join(ctx.RootForImplicitAbsolutePaths, systemRegistriesConfPath)
+		} else {
+			wrapper.configPath = systemRegistriesConfPath
+		}
 	}
+
+	if ctx != nil && ctx.SystemRegistriesConfDirPath != "" {
+		wrapper.configDirPath = ctx.SystemRegistriesConfDirPath
+	} else if ctx != nil && ctx.RootForImplicitAbsolutePaths != "" {
+		wrapper.configDirPath = filepath.Join(ctx.RootForImplicitAbsolutePaths, systemRegistriesConfDirPath)
+		wrapper.userConfigDirPath = userRegistriesDirPath
+	} else {
+		wrapper.configDirPath = systemRegistriesConfDirPath
+		wrapper.userConfigDirPath = userRegistriesDirPath
+	}
+
+	return wrapper
+}
+
+// ConfigurationSourceDescription returns a string containres paths of registries.conf and registries.conf.d
+func ConfigurationSourceDescription(ctx *types.SystemContext) string {
+	wrapper := newConfigWrapper(ctx)
+	return strings.Join([]string{wrapper.configPath, wrapper.configDirPath, wrapper.userConfigDirPath}, ", ")
```

Comment:
I don’t think this does the right thing if either of the `*DirPath` values is `""`.